### PR TITLE
 Membership signup form

### LINKED ISSUE #157: Membership Signup
image

### PR DESCRIPTION
**Pull request by LouiaR**
Fixes #157

### CODE REVIEW

File: src/components/Molecules/Membership/Form/Form.js [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+    setMoneyBuyCopy(copy);
+    setAmountDonate(value);
+  };
+
+  // Handle user input
+  const handleChange = (input, otherDescription) => {
+    if (!isAmountValid(input)) {
+      setMoneyBuyCopy(false);
+      setErrorMsg(true);
+    } else if (input * 1 === box1.amount) {
```
[the remark this answers is not in the document]
Comment (LouiaR) [reply src/components/Molecules/Membership/Form/Form.js]: Sure

File: src/components/Molecules/Membership/Form/Form.js [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+  const [userInput, setUserInput] = useState('');
+  const [boxBorderColor, setBoxBorderColor] = useState('');
+  const [inputBorderColor, setInputBorderColor] = useState(false);
+  const [moneyBuyCopy, setMoneyBuyCopy] = useState(true);
+
+  const selectMoneyBuy = (copy, value) => {
+    setIsSelected(false);
+    setErrorMsg(false);
+    setAmountDonate('');
+    if (inputBorderColor && userInput * 1 !== value * 1) {
```
[the remark this answers is not in the document]
Comment (LouiaR) [reply src/components/Molecules/Membership/Form/Form.js]: Sure

File: src/components/Molecules/Membership/Form/Form.js [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+      setBoxBorderColor(false);
+      setErrorMsg(false);
+      setMoneyBuyCopy(otherDescription);
+      setAmountDonate(input);
+    }
+    setInputBorderColor(true);
+    setUserInput(input);
+  };
+
+  const hightlightInput = (value, otherDescription) => {
```
[the remark this answers is not in the document]
Comment (LouiaR) [reply src/components/Molecules/Membership/Form/Form.js]: Will give a try

File: src/components/Molecules/Membership/Form/Form.js [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+  const [amountDonate, setAmountDonate] = useState('');
+  const [userInput, setUserInput] = useState('');
+  const [boxBorderColor, setBoxBorderColor] = useState('');
+  const [inputBorderColor, setInputBorderColor] = useState(false);
+  const [moneyBuyCopy, setMoneyBuyCopy] = useState(true);
+
+  const selectMoneyBuy = (copy, value) => {
+    setIsSelected(false);
+    if (errorMsg) setErrorMsg(false);
+    const isUserInputMatch = userInput * 1 === value * 1;
```
Comment (gusliedke) [inline src/components/Molecules/Membership/Form/Form.js]: `userInput === value` is the same thing, no?!

File: src/components/Molecules/Membership/Form/Form.js [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+      );
+    });
+  }, [isSelected, regularGiving.moneybuys]);
+
+  const boxes = regularGiving.moneybuys.map(({ value, description }, index) => (
+    <MoneyBuy
+      isSelected={index === 1 && isSelected}
+      boxBorderColor={boxBorderColor}
+      current={value}
+      amount={`${value}`}
```
Comment (gusliedke) [inline src/components/Molecules/Membership/Form/Form.js]: `{value}` same thing?

File: src/components/Molecules/Membership/Membership.style.js [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+const FormWrapper = styled.div`
+  box-shadow: 0px ${spacing('md')} ${spacing('xl')} rgba(0, 0, 0, 0.3);
+  height: 450px;
+
+  ${media('small')} {
+    height: 430px;
+  }
+`;
+
+const Error = styled(Text)`
```
Comment (gusliedke) [inline src/components/Molecules/Membership/Membership.style.js]: Can we use Text component directly?

File: src/components/Molecules/Membership/Membership.style.js [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+const AmountField = styled(Input)`
+  position: relative;
+  flex: 0 0 50%;
+  font-weight: 400;
+
+  ${media('small')} {
+    flex: 0 0 60%;
+  }
+
+  span {
```
Comment (gusliedke) [inline src/components/Molecules/Membership/Membership.style.js]: Can't we use FlexBox to align item?

File: src/components/Molecules/Membership/MoneyBuy/MoneyBuy.js [hunk 8]
```diff
@@ -0,0 +1,59 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import styled, { css } from 'styled-components';
+
+import Input from '../../../Atoms/Input/Input';
+
+const MoneyBox = styled(Input)`
+  display: block;
+  input {
```
Comment (gusliedke) [inline src/components/Molecules/Membership/MoneyBuy/MoneyBuy.js]: Input inside Input?

File: src/components/Molecules/Membership/Form/Form.js (hunk 4 above)
Comment (LouiaR) [reply src/components/Molecules/Membership/Form/Form.js]: Nope `userInput` is a string

**Comment by gusliedke** [comment]
Isn't worth using [Formik](https://jaredpalmer.com/formik/docs/overview) to deal with all form validations, etc

File: src/components/Molecules/Membership/Membership.style.js (hunk 6 above)
Comment (LouiaR) [reply src/components/Molecules/Membership/Membership.style.js]: I may add this icon later, or is it ok to display error message without it
Screen Shot 2019-10-09 at 11 39 34

File: src/components/Molecules/Membership/MoneyBuy/MoneyBuy.js (hunk 8 above)
Comment (LouiaR) [reply src/components/Molecules/Membership/MoneyBuy/MoneyBuy.js]: No, input is wrapped inside `label` tag

File: src/components/Molecules/Membership/Form/Form.js (hunk 5 above)
Comment (LouiaR) [reply src/components/Molecules/Membership/Form/Form.js]: Props validation was set to string will set it to number

File: src/components/Molecules/Membership/Form/Form.js (hunk 4 above)
Comment (gusliedke) [reply src/components/Molecules/Membership/Form/Form.js]: I mean `userInput === value` is the same as `userInput * 1 === value * 1`. No? Why multiply a string by 1?

File: src/components/Molecules/Membership/Membership.style.js (hunk 6 above)
Comment (gusliedke) [reply src/components/Molecules/Membership/Membership.style.js]: later we can add the Icon

File: src/components/Molecules/Membership/Form/Form.js (hunk 4 above)
Comment (LouiaR) [reply src/components/Molecules/Membership/Form/Form.js]: Sorry,  it should be like this `userInput * 1 === value`

**Comment by AndyEPhipps** [comment]
Can we change the focus state border to 2px? Stops it the squirming around:

focusgif

@LouiaR - I've also noticed that we're not handling the Monthly-Single functionality at all?

File: src/utils/Membership.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+  const affiliateValue = getUrlParameter(window.location.href);
+  const currentpageUrl = window.location.href;
+  const urlString =
+    currentpageUrl.indexOf('?') > -1
+      ? currentpageUrl.substring(0, currentpageUrl.indexOf('?'))
+      : currentpageUrl;
+  // window.location.href = (donationLink + "?clientOverride=" + clientId + "&amount=" + amount + "&currency=" + currency + "&givingType=" + givingType + "&cartId=" + cartId + "&affiliate=" + affiliateValue + "&siteurl=" + url_string + '&rowID=' + rowID + '&moneybuy=' + submitNameID
+  console.log(urlString, affiliateValue, amount);
+  window.location.href = `${donateLink}?clientOverride=67888&amount=${amount}&currency=£&givingType=monthly&cartId=00000&affiliate=${affiliateValue}&siteurl=${urlString}`;
+};
```
Comment (AndyEPhipps) [inline src/utils/Membership.js]: We'll need to set this Monthly or Single value too

File: src/utils/Membership.js [hunk 10]
```diff
@@ -0,0 +1,53 @@
+const onKeyPress = event => {
+  const keyCode = event.keyCode || event.which;
+  const keyValue = String.fromCharCode(keyCode);
+  if (/\+|-|e/.test(keyValue)) event.preventDefault();
+};
```
Comment (AndyEPhipps) [inline src/utils/Membership.js]: What is this function doing? Can we add some comments?

File: src/utils/Membership.js [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+  const param =
+    results === null
+      ? 'generic'
+      : decodeURIComponent(results[1].replace(/\+/g, ' '));
+  return param;
+};
+
+const donate = (e, amount) => {
+  e.preventDefault();
+  const donateLink = 'https://donation.comicrelief.com/';
```
Comment (AndyEPhipps) [inline src/utils/Membership.js]: Can we set this via a prop?

Also, would be nice to rename the function to handleSubmission or something more standard and identifiable

**Comment by LouiaR** [comment]
> Can we change the focus state border to 2px? Stops it the squirming around:
> 
> focusgif
> 
> @LouiaR - I've also noticed that we're not handling the Monthly-Single functionality at all?

Yeah, we agreed with @gusliedke to focus on monthly first.

File: src/utils/Membership.js [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+  const affiliateValue = getUrlParameter(window.location.href);
+  const currentpageUrl = window.location.href;
+  const urlString =
+    currentpageUrl.indexOf('?') > -1
+      ? currentpageUrl.substring(0, currentpageUrl.indexOf('?'))
+      : currentpageUrl;
+  // window.location.href = (donationLink + "?clientOverride=" + clientId + "&amount=" + amount + "&currency=" + currency + "&givingType=" + givingType + "&cartId=" + cartId + "&affiliate=" + affiliateValue + "&siteurl=" + url_string + '&rowID=' + rowID + '&moneybuy=' + submitNameID
```
Comment (AndyEPhipps) [inline src/utils/Membership.js]: I guess we can get rid of this now?

**Comment by AndyEPhipps** [comment]
> Yeah, we agreed with @gusliedke to focus on monthly first.

Cool 👍

File: src/components/Molecules/Membership/MoneyBuy/MoneyBuy.js (hunk 8 above)
Comment (gusliedke) [reply src/components/Molecules/Membership/MoneyBuy/MoneyBuy.js]: ok Got it

File: src/components/Molecules/Membership/Form/Form.js (hunk 4 above)
Comment (AndyEPhipps) [reply src/components/Molecules/Membership/Form/Form.js]: Could always use parseInt to make it clearer what's going on

File: src/components/Molecules/Membership/Form/Form.js [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+  };
+
+  useEffect(() => {
+    regularGiving.moneybuys.map((moneyBuy, index) => {
+      const box = `box${index + 1}`;
+      // eslint-disable-next-line no-shadow
+      setMoneyBoxes(moneyBoxes => ({ ...moneyBoxes, [box]: moneyBuy }));
+      return (
+        isSelected &&
+        index === 1 &&
```
Comment (lebaz20) [inline src/components/Molecules/Membership/Form/Form.js]: I believe `index === 1` that what makes the second box, selected by default, i am not sure if there's no flexibility required here, for instance, if we wanted more than three boxes, or default one to be something other than the second one, is that an actual concern ?

File: src/utils/Membership.js [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+      moneyBoxes[box].value === inputToNum &&
+      (selectMoneyBuy(moneyBoxes[box].description, moneyBoxes[box].value),
+      setAmountDonate(input))
+    );
+  });
+};
+
+// This function is still in progress
+const handleDonateSubmission = (amount, clientID, cartID, mbshipID) => {
+  const donateLink = 'https://donation.comicrelief.com/';
```
Comment (lebaz20) [inline src/utils/Membership.js]: this needs to be customizable, in development, for instance, we will need to test on another url, and if another domain to be used as well, like sport relief.

File: src/components/Molecules/Membership/Form/Form.js (hunk 13 above)
Comment (LouiaR) [reply src/components/Molecules/Membership/Form/Form.js]: Yep, it selects the second box, yeah maybe later if we plan to add more boxes we could set field (bool) within contentful to control that... or something better

File: src/utils/Membership.js (hunk 14 above)
Comment (LouiaR) [reply src/utils/Membership.js]: Sure

File: src/components/Molecules/Membership/Membership.js [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+const Membership = ({
+  cartID,
+  backgroundColor,
+  data,
+  title,
+  subtitle,
+  otherDescription,
+  formAligntRight,
+  vhFull,
+  fullImage,
```
[the remark this answers is not in the document]
Comment (LouiaR) [reply src/components/Molecules/Membership/Membership.js]: Yeah, I remove it

File: src/utils/Membership.js [hunk 16]
```diff
@@ -0,0 +1,58 @@
+// this function prevent keyboard characters like  e, + , - to be pass on the input
+const onKeyPress = event => {
+  const keyCode = event.keyCode || event.which;
+  const keyValue = String.fromCharCode(keyCode);
+  if (/\+|-|e/.test(keyValue)) event.preventDefault();
+};
+
+const isAmountValid = input => {
+  const reg = /^\s*(?=.*[1-9])\d*(?:\.\d{1,2})?\s*$/g;
```
[the remark this answers is not in the document]
Comment (LouiaR) [reply src/utils/Membership.js]: Yeah, it's to check that I have a valid number with 2 decimal place
Validation happens on the next line

File: src/utils/Membership.js (hunk 14 above)
Comment (LouiaR) [reply src/utils/Membership.js]: Done

**Comment by AndyEPhipps** [comment]
Also, let's add in the right-aligned version, just so we're able to check all variations in CL 👍

**Comment by AndyEPhipps** [comment]
Screen Shot 2019-10-17 at 09 54 40

Screen Shot 2019-10-17 at 09 54 40

Let's see if we can match that alignment too plzzz

File: src/components/Atoms/Input/input.test.js [hunk 17]
```diff
@@ -80,12 +80,12 @@ it('renders correctly', () => {
 
     <label
       className="c0"
-      htmlFor="Acessibility info go here"
+      htmlFor="undefined"
```
Comment (gusliedke) [inline src/components/Atoms/Input/input.test.js]: Undefined? Why all the changes in the snaps?

Comment (AndyEPhipps) [reply src/components/Atoms/Input/input.test.js]: I just the same from pulling master, not sure what's going on

Comment (LouiaR) [reply src/components/Atoms/Input/input.test.js]: I pulled the master

Comment (gusliedke) [reply src/components/Atoms/Input/input.test.js]: `lazyload c5` to `c5 lazyload` and `C2` to `C1` are fine but `htmlFor="undefined"` is a bit worrying .. I will reinstall modules and run test in master before we merge those changes in all PRs.

File: src/utils/Membership.js (hunk 10 above)
Comment (LouiaR) [reply src/utils/Membership.js]: Done

**Comment by AndyEPhipps** [comment]
Looking good @LouiaR; just need to get that 'scroll-to' functionality in.

I think it makes the most sense to handle it here (rather than within Contentful) as you'll be able to hook into the existing functions that run on ComponentDidMount. If we keep it within Contentful, we'd need to repeat the scroll code in any other projects that use the Row.

**Comment by AndyEPhipps** [comment]
I guess the mobile image doesn't need to be its own separate prop, we can just pass it from Contentful via `images`?

**Comment by LouiaR** [comment]
Hey @AndyEPhipps, I think a new ticket will handle : 
background image mobile and scrollTo functionality as we need also to highlight moneybox previously selected or prefill other amount field input with the right amount previously defined by the user.
Let me know what you think.

**Comment by AndyEPhipps** [comment]
> Hey @AndyEPhipps, I think a new ticket will handle :
> background image mobile and scrollTo functionality as we need also to highlight moneybox previously selected or prefill other amount field input with the right amount previously defined by the user.
> Let me know what you think.

Sure, Idon't see why not- would you be able to pick that up straight after getting this into Contentful, so we don't forget about it?

**Comment by LouiaR** [comment]
> > Hey @AndyEPhipps, I think a new ticket will handle :
> > background image mobile and scrollTo functionality as we need also to highlight moneybox previously selected or prefill other amount field input with the right amount previously defined by the user.
> > Let me know what you think.
> 
> Sure, Idon't see why not- would you be able to pick that up straight after getting this into Contentful, so we don't forget about it?

Sure

File: yarn.lock [hunk 18]
```diff
@@ -2627,6 +2627,16 @@ browserify-zlib@^0.2.0:
   dependencies:
     pako "~1.0.5"
 
+<<<<<<< HEAD
+browserslist@4.7.0, browserslist@^4.0.0, browserslist@^4.1.1, browserslist@^4.4.2, browserslist@^4.6.0, browserslist@^4.7.0:
```
Comment (AndyEPhipps) [inline yarn.lock]: Make sure you're always recreating lockfiles after pulling in master

Comment (LouiaR) [reply yarn.lock]: Done

**Pull request merged**